homeassistant/components/rainbird/__init__.py

```diff
@@ -1,42 +1,88 @@
+)
+from homeassistant.helpers import discovery
 import homeassistant.helpers.config_validation as cv
-from homeassistant.const import CONF_HOST, CONF_PASSWORD
+
+CONF_ZONES = "zones"
+
+SUPPORTED_PLATFORMS = [switch.DOMAIN, sensor.DOMAIN, binary_sensor.DOMAIN]
 
 _LOGGER = logging.getLogger(__name__)
 
+RAINBIRD_CONTROLLER = "controller"
 DATA_RAINBIRD = "rainbird"
 DOMAIN = "rainbird"
 
-CONFIG_SCHEMA = vol.Schema(
+SENSOR_TYPE_RAINDELAY = "raindelay"
+SENSOR_TYPE_RAINSENSOR = "rainsensor"
+# sensor_type [ description, unit, icon ]
+SENSOR_TYPES = {
+    SENSOR_TYPE_RAINSENSOR: ["Rainsensor", None, "mdi:water"],
+    SENSOR_TYPE_RAINDELAY: ["Raindelay", None, "mdi:water-off"],
+}
+
+ZONE_SCHEMA = vol.Schema(
     {
-        DOMAIN: vol.Schema(
-            {vol.Required(CONF_HOST): cv.string, vol.Required(CONF_PASSWORD): cv.string}
-        )
-    },
+        vol.Optional(CONF_FRIENDLY_NAME): cv.string,
+        vol.Optional(CONF_TRIGGER_TIME): cv.positive_int,
+    }
+)
+CONTROLLER_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_TRIGGER_TIME): cv.string,
+        vol.Optional(CONF_ZONES): vol.Schema({cv.positive_int: ZONE_SCHEMA}),
+    }
+)
+CONFIG_SCHEMA = vol.Schema(
+    {DOMAIN: vol.Schema(vol.All(cv.ensure_list, [CONTROLLER_SCHEMA]))},
     extra=vol.ALLOW_EXTRA,
 )
 
 
 def setup(hass, config):
     """Set up the Rain Bird component."""
-    conf = config[DOMAIN]
-    server = conf.get(CONF_HOST)
-    password = conf.get(CONF_PASSWORD)
 
-    from pyrainbird import RainbirdController
+    hass.data[DATA_RAINBIRD] = []
+    success = False
+    for controller_config in config[DOMAIN]:
+        success = success or _setup_controller(hass, controller_config)
 
-    controller = RainbirdController(server, password)
+    return success
 
-    _LOGGER.debug("Rain Bird Controller set to: %s", server)
 
-    initial_status = controller.currentIrrigation()
-    if initial_status and initial_status["type"] != "CurrentStationsActiveResponse":
-        _LOGGER.error("Error getting state. Possible configuration issues")
-        return False
+def _setup_controller(hass, config):
 
-    hass.data[DATA_RAINBIRD] = controller
-    return True
+    server = config[CONF_HOST]
+    password = config[CONF_PASSWORD]
+    controller = RainbirdController(server, password)
+    position = len(hass.data[DATA_RAINBIRD])
+    success = False
+    try:
+        controller.get_serial_number()
+        hass.data[DATA_RAINBIRD].append(controller)
+        _LOGGER.debug("Rain Bird Controller %d set to: %s", position, server)
+        for platform in SUPPORTED_PLATFORMS:
+            discovery.load_platform(
+                hass,
+                platform,
+                DOMAIN,
+                discovered={**{RAINBIRD_CONTROLLER: position}, **config},
+                hass_config=config,
+            )
+        success = True
+    except Exception as e:
```

Comment:
Please only wrap the call that can error with try... except.

We should try to except specific exceptions. What error do we expect?

Reply:
We don't want to initialize platform when controller is unreachable. That's why also loop is inside the exception.

There is no specific exception. Just `Exception`

Reply:
Please change like this:
```py
def _setup_controller(hass, controller_config, config):
    """Set up a controller."""
    server = controller_config[CONF_HOST]
    password = controller_config[CONF_PASSWORD]
    controller = RainbirdController(server, password)
    position = len(hass.data[DATA_RAINBIRD])
    try:
        controller.get_serial_number()
    except Exception as exc:
        _LOGGER.error("Unable to setup controller: %s", exc)
        return False
    hass.data[DATA_RAINBIRD].append(controller)
    _LOGGER.debug("Rain Bird Controller %d set to: %s", position, server)
    for platform in SUPPORTED_PLATFORMS:
        discovery.load_platform(
            hass, platform, DOMAIN, {RAINBIRD_CONTROLLER: position, **controller_config}, config,
        )
    return True
```

The library should preferably be updated to raise library specific exceptions. It's not good to raise general exceptions like `Exception` directly. Create a library base exception that inherits from `Exception`. Then create further library exceptions as needed that all inherit from the base library exception.

Reply:
done